File: MatlabCode/ChapterTISE/Exercise_3_4_3.m
% This script determines the ground state for a specific potential 
% by means of propagation in inmaginary time.
% The potential barrier has a "smoothly rectangular" shape.
%
% Numerical inputs:
%   L       - The extension of the spatial grid 
%   N       - The number of grid points
%   Tfinal  - The duration of the simulation - in imaginary time
%   dt      - The step size in imaginary time
%
% Input for the potential:
%   V0      - The "height" of the barrier (must be negative)
%   w       - The width of the barrier
%   s       - Smoothness parameter
%   Vpot    - The potential (function variable)
%
% All inputs are hard coded initially.

% Numerical grid parameters
L = 20;
N = 512;                % Should be 2^k, k integer, for FFT's sake
h = L/(N-1);

% Numerical time parameters
Tfinal = 10;
dt = 1e-2;

% Input parameters for the barrier
V0 = -3;
w = 8;
s = 5;

% The potential
Vpot = @(x) V0./(exp(s*(abs(x)-w/2))+1);

% Set up grid
x = transpose(linspace(-L/2, L/2, N));      % Column vector

% Set up kinetic energy matrix by means of the fast Fourier transform
k=2*pi/(N*h)*[0:(N/2-1), (-N/2:-1)];          % Vector with k-values
% Fourier transform the identity matrix:
Tmat_FFT = fft(eye(N));
% Multiply by (ik)^2
Tmat_FFT = diag(-k.^2)*Tmat_FFT;
% Transform back to x-representation
Tmat_FFT = ifft(Tmat_FFT);
Tmat_FFT = -1/2*Tmat_FFT;            % Correct prefactor


% Set up propagator for the approximation
Ham = Tmat_FFT + diag(Vpot(x));         % Hamiltonian
U = expm(-Ham*dt);                      % Propagator

% Set up initial state - just a random "wave function"
Psi0 = rand(N, 1);
InitialNorm = trapz(x, abs(Psi0).^2);
% Impose proper norm
Psi0 = Psi0/sqrt(InitialNorm);

% Initiate plots
figure(1)
pl = plot(x,abs(Psi0).^2, 'k-', 'linewidth', 1.5);
set(gca, 'fontsize', 15)
xlabel('x')
ylabel('|\Psi|^2')

% Initiate wave functons and "time"
Psi = Psi0;
t = 0;

% Initiate/allocate vectors for plotting
TimeVector = 0:dt:Tfinal;
EnergyVector = zeros(1, length(TimeVector));

% Loop which updates wave functions and plots in time
index = 1;
for t = TimeVector  
  % Update time
  t = t+dt;             
  % Update numerical approximation
  Psi = U*Psi;
  
  % Normalize again
  N2 = trapz(x, abs(Psi).^2);
  Energy = -1/(2*dt)*log(N2);
  EnergyVector(index) = Energy;  % Energy estimate
  Psi = Psi/sqrt(N2);
  
  % Update plot
  set(pl, 'ydata', abs(Psi).^2);
  drawnow
  
  % Update index
  index = index + 1;
end


% Determine actual (numerical) ground state
% Diagonalize to find eigen states and energies
[EigVectors EigValues] = eig(Ham);
% Extract eigen values from diagonal matrix
EigValues = diag(EigValues);
% Sort them
[EigValues, Indexes] = sort(real(EigValues));
EigVectors = EigVectors(:,Indexes);
% "Exact" groundState
GroundStateEnergy = EigValues(1);
GroundStateWF = EigVectors(:,1)/sqrt(h);  % Normalize

% Plot energy as a function of "time"
figure(2)
plot(TimeVector, EnergyVector, 'k-', 'linewidth', 2)
hold on
yline(GroundStateEnergy, 'r--', 'linewidth', 2);
hold off
grid on
set(gca, 'fontsize', 15)
xlabel('Imaginary time')
ylabel('Energy')
legend('Estimate', 'True ground state energy')

% Compare wave functions
figure(3)
plot(x, abs(Psi).^2, 'k-', 'linewidth', 2)
hold on
% Plott "exact" wave function
plot(x, abs(GroundStateWF).^2, 'r--', 'linewidth', 2)          
% Fix strength of the potential for plotting
MaxWalPot = -.2*max(abs(Psi)).^2/V0;          
% Plot confining potential
plot(x, Vpot(x)*MaxWalPot, 'b');      
hold off
grid on
set(gca,'fontsize',15)
legend('Estimate','Exact','Potential')

% Write estimates to screen
disp(['Energy from imaginary time: ', num2str(Energy)])
disp(['Energy from direct diagonalization: ', num2str(GroundStateEnergy)])

File: PythonCode/Chapter3_Time-indepdentSE/Exercise_3_4_3.py
"""
 This script determines the ground state for a specific 
 potential by means of propagation in inmaginary time.
 The potential barrier has a "smoothly rectangular" shape.

 Numerical inputs:
   L       - The extension of the spatial grid 
   N       - The number of grid points
   Tfinal  - The duration of the simulation - in imaginary time
   dt      - The step size in imaginary time

 Input for the potential:
   V0      - The "height" of the barrier (must be negative)
   w       - The width of the barrier
   s       - Smoothness parameter
   Vpot    - The confining potential (function variable)

 All inputs are hard coded initially.
"""

# Import libraries  
import numpy as np
from matplotlib import pyplot as plt
from scipy import linalg

# Numerical grid parameters
L = 20.0
N = 256         # Should be 2**k, with k being an integer

# Numerical time parameters
Tfinal = 10;
dt = 0.05;

# Inputs for the smoothly rectangular potential
V0 = -3.0            # Heigth
width = 8.0         # Width
s = 5.0             # Smoothness

# The potential
def Vpot(x):
    return V0/(np.exp(s*(np.abs(x)-width/2))+1)

# Set up grid
x = np.linspace(-L/2, L/2, N)
h = L/(N-1)

# Determine double derivative by means of the 
# fast Fourier transform.
# Set up vector of k-values
k_max = np.pi/h
dk = 2*k_max/N
k = np.append(np.linspace(0, k_max-dk, int(N/2)), 
              np.linspace(-k_max, -dk, int(N/2)))
# Transform identity matrix
Tmat_FFT = np.fft.fft(np.identity(N, dtype=complex), axis = 0)
# Multiply by (ik)^2
Tmat_FFT = np.matmul(np.diag(-k**2), Tmat_FFT)
# Transform back to x-representation. 
Tmat_FFT = np.fft.ifft(Tmat_FFT, axis = 0)
# Correct pre-factor
Tmat_FFT = -1/2*Tmat_FFT    

# Full Hamiltonian
Ham = Tmat_FFT + np.diag(Vpot(x))
# Propagator
U = linalg.expm(-Ham*dt)

# Initiate random wave functons and time
Psi0 = np.random.rand(N)
NormInit = np.trapz(np.abs(Psi0)**2, dx = h)
Psi0 = Psi0/np.sqrt(NormInit)                 
Psi0 = Psi0.reshape(N,1)

# Initiate plots
plt.ion()
fig = plt.figure(1)
plt.clf()
ax = fig.add_subplot()
line1, = ax.plot(x, np.abs(Psi0)**2, '-', color='black')
plt.xlabel('x', fontsize = 12)
plt.ylabel(r'$|\Psi|^2$', fontsize = 12)

# Initiate wave function
Psi = Psi0
# Vectors for plotting
Ntime = int(Tfinal/dt)
tVector = np.linspace(0, Tfinal, Ntime)
EnergyVector = 0*tVector

# Iterate while time t is less than final time Tfinal
index = 0
for t in tVector:
  # Update wave function
  Psi = np.matmul(U, Psi)
  # Normalize
  Norm2 = float(np.trapz(np.abs(Psi.T)**2, x))
  Psi = Psi/np.sqrt(Norm2)
  # Energy estimate
  EnergyEstimate = -0.5/dt*np.log(Norm2)
  EnergyVector[index] = EnergyEstimate
  # Update data for plots
  line1.set_ydata(np.power(np.abs(Psi), 2))
  # recompute the ax.dataLim
  ax.relim()    
  # update ax.viewLim using the new dataLim
  ax.autoscale_view()  # Update plots
  # Adjust axes
  fig.canvas.draw()
  #fig.canvas.flush_events()
  plt.pause(0.01)
  # Update index
  index = index + 1
  
  
# Determine actual enegy eigen state  
Evector, PsiMat = np.linalg.eigh(Ham)
# Extract ground state
GroundStateEnergy = Evector[0]

  
# Plot "evolution" of energy estimate
plt.figure(2)
plt.clf()
plt.plot(tVector, EnergyVector, label = 'Estimated energy')  
plt.axhline(y = GroundStateEnergy, color = 'r', 
            linestyle = '--', label = 'Actual energy')
plt.grid()
plt.legend()
plt.xlabel('Imaginary time', fontsize = 12)
plt.ylabel('Energy estimate', fontsize = 12)
plt.show()

# Diagonalize Hamiltonian
Evector, PsiMat = np.linalg.eigh(Ham)
# Extract ground state
GroundStateEnergy = Evector[0]
GroundState = PsiMat[:,0]
# Normalize
GroundState = PsiMat[:,0]
GroundState = GroundState/np.sqrt(h)

# Plot comparison with the ground state from direct diagonalization
plt.figure(3)
plt.clf()
plt.plot(x, np.abs(Psi)**2, 'k-', label = 'Imaginary time')
# Actual
plt.plot(x, np.abs(GroundState)**2, 'r--', label = 'Diagonalization')
# Potential
ScaleFactor = np.max(np.abs(Psi)**2)
ScaleFactor = 0.2*ScaleFactor/np.abs(V0)
plt.plot(x, ScaleFactor*Vpot(x), 'b', label = 'Potential')
# Cosmetics
plt.xlabel('x', fontsize = 12)
plt.legend()
plt.grid()
plt.show()

# Write estimate to screen
print(f'Estimated ground state energy: {EnergyEstimate:.4f}')
print(f'Actual ground state energy on the grid: {GroundStateEnergy:.4f}')